### PR TITLE
feat: modify ballot page styling

### PR DESCRIPTION
**Pull request by kittybest**
**Description**
- [x] add identifier to choose what to display on sidebar
- [x] add submit button
- [x] edit ballot allocation item styling

**Blocker**
#164 ✅

### CODE REVIEW

File: src/features/ballot/components/AllocationList.tsx [hunk 1]
```diff
@@ -79,91 +54,50 @@ export function AllocationFormWrapper({
...
-      <Table>
-        {renderHeader?.()}
-        <Tbody>
-          {fields.map((project, i) => {
-            const idx = i;
+    <Table>
+      {renderHeader?.()}
+      <Tbody>
+        {fields.map((project, i) => {
+          const idx = i;
```
Comment (ctrlc03) [inline src/features/ballot/components/AllocationList.tsx:61]: why not call it directly idx on line 60?

File: src/features/ballot/components/BallotConfirmation.tsx [hunk 2]
```diff
@@ -1,88 +1,136 @@
...
-      </div>
+      )}
+      <Card>
+        <div className="flex-3 flex flex-col gap-4">
+          <b className="font-mono text-2xl uppercase">
+            Help us improve our next round of {config.eventName}
+          </b>
+          <p className="text-gray-400">
+            Your anonymized feedback will be influential to help us iterate on
+            Optimism’s RetroPGF process.
```
Comment (ctrlc03) [inline src/features/ballot/components/BallotConfirmation.tsx:100]: should we remove Optimism mention here?

Comment (ctrlc03) [reply src/features/ballot/components/BallotConfirmation.tsx:100]: also, are we actually saving any feedback?

File: src/features/ballot/components/AllocationList.tsx (hunk 1 above)
Comment (kittybest) [reply src/features/ballot/components/AllocationList.tsx:61]: good question, I just don't modify it, let me modify it

File: src/features/ballot/components/BallotConfirmation.tsx (hunk 2 above)
Comment (kittybest) [reply src/features/ballot/components/BallotConfirmation.tsx:100]: currently we just link the feedback to create a github issue, maybe should create a google form for it?

File: package.json [hunk 3]
```diff
@@ -52,6 +52,7 @@
     "react-dom": "18.2.0",
     "react-hook-form": "^7.49.3",
     "react-icons": "^5.0.1",
+    "react-loading-animation": "^1.4.0",
```
Comment (0xmad) [inline package.json:55]: This package was published 7 years ago and I don't think we need loading components as external dependency

File: src/components/Header.tsx [hunk 4]
```diff
@@ -49,6 +51,11 @@ export const Header = ({ navLinks }: { navLinks: NavLink[] }) => {
               href={link.href}
             >
               {link.children}
+              {link.href === "/ballot" && ballot && ballot.votes.length > 0 && (
+                <div className="ml-2 h-5 w-5 rounded-full border-2 border-blue-400 bg-blue-50 text-center text-sm leading-4 text-blue-400">
+                  {ballot?.votes.length}
```
Comment (0xmad) [inline src/components/Header.tsx:56]:
```suggestion
                  {ballot.votes.length}
```

File: src/components/ui/Notification.tsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    },
+  },
+  defaultVariants: {
+    variant: "default",
+  },
+});
+
+const NotificationContainer = createComponent("div", notification);
+
+type NotificationProps = {
```
Comment (0xmad) [inline src/components/ui/Notification.tsx:22]: Use interface instead

File: src/contexts/Ballot.tsx [hunk 6]
```diff
@@ -75,6 +75,11 @@ export const BallotProvider: React.FC<BallotProviderProps> = ({ children }) => {
 
   /// Read existing ballot in localStorage
   useEffect(() => {
+    const defaultBallot = {
+      votes: [],
+      published: false,
+    };
```
Comment (0xmad) [inline src/contexts/Ballot.tsx:81]: If you add this const, you need to remove the global one

File: src/features/ballot/components/ProjectAvatarWithName.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
+import Link from "next/link";
+
+import { ProjectAvatar } from "~/features/projects/components/ProjectAvatar";
+import {
+  useProjectById,
+  useProjectMetadata,
+} from "~/features/projects/hooks/useProjects";
+
+type Props = {
```
Comment (0xmad) [inline src/features/ballot/components/ProjectAvatarWithName.tsx:9]: Interface and change name from this generic one

File: src/features/ballot/components/SubmitBallotButton.tsx [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+import { Dialog } from "~/components/ui/Dialog";
+
+export const SubmitBallotButton = () => {
+  const router = useRouter();
+  const [isOpen, setOpen] = useState(false);
+  const { onVote, isLoading } = useMaci();
+  const { ballot, publishBallot } = useBallot();
+  const projectIndices = useProjectIdMapping(ballot);
+
+  const handleSubmitBallot = async () => {
```
Comment (0xmad) [inline src/features/ballot/components/SubmitBallotButton.tsx:17]: useCallback

File: src/layouts/DefaultLayout.tsx [hunk 9]
```diff
@@ -19,17 +22,33 @@ type Props = PropsWithChildren<
...
-      href: "/projects",
-      children: "Projects",
-    },
-    {
-      href: "/ballot",
-      children: "My Ballot",
-    },
-  ];
+  const navLinks = useMemo(() => {
+    if (ballot && ballot.published) {
```
Comment (0xmad) [inline src/layouts/DefaultLayout.tsx:28]:
```suggestion
    if (ballot?.published) {
```

File: src/pages/ballot/confirmation.tsx [hunk 10]
```diff
@@ -1,17 +1,29 @@
...
 import { Layout } from "~/layouts/DefaultLayout";
 
 export default function BallotConfirmationPage() {
+  const [isLoading, setIsLoading] = useState<boolean>(true);
+
   const { ballot } = useBallot();
+  const router = useRouter();
+
+  useEffect(() => {
+    if (ballot && ballot.published) {
```
Comment (0xmad) [inline src/pages/ballot/confirmation.tsx:15]:
```suggestion
    if (ballot?.published) {
```

**Review by 0xmad — CHANGES_REQUESTED** [review]
@kittybest thanks, just left some comments

File: pnpm-lock.yaml
Comment (0xmad) [inline pnpm-lock.yaml:1]: If there are no package.json changes, lock file should not be changes as well

File: src/components/ui/Loading.tsx
Comment (0xmad) [inline src/components/ui/Loading.tsx:1]: Where do you use this component?

File: src/features/ballot/components/SubmitBallotButton.tsx [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+    await onVote(
+      votes,
+      async () => {},
+      async () => {
+        publishBallot();
+        await router.push("/ballot/confirmation");
+      },
+    );
+  }, [ballot, onVote, publishBallot]);
```
Comment (0xmad) [inline src/features/ballot/components/SubmitBallotButton.tsx:32]:
```suggestion
  }, [ballot, router, onVote, publishBallot]);
```

File: src/features/ballot/components/SubmitBallotButton.tsx [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+  const handleSubmitBallot = useCallback(async () => {
+    const votes =
+      ballot?.votes.map(({ amount, projectId }) => ({
+        voteOptionIndex: BigInt(projectIndices[projectId]),
+        newVoteWeight: BigInt(amount),
+      })) ?? [];
+
+    await onVote(
+      votes,
+      async () => {},
```
Comment (0xmad) [inline src/features/ballot/components/SubmitBallotButton.tsx:26]: It should have some logic for error handling or be undefined to have default handler

File: src/features/ballot/components/SubmitBallotButton.tsx [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+      async () => {
+        publishBallot();
+        await router.push("/ballot/confirmation");
+      },
+    );
+  }, [ballot, onVote, publishBallot]);
+
+  return (
+    <>
+      <Button variant="primary" onClick={() => setOpen(true)}>
```
Comment (0xmad) [inline src/features/ballot/components/SubmitBallotButton.tsx:36]: Click handler should have a callback declaration like onVote.

File: src/pages/ballot/confirmation.tsx [hunk 14]
```diff
@@ -1,17 +1,29 @@
...
+  const router = useRouter();
+
+  useEffect(() => {
+    if (ballot?.published) {
+      setIsLoading(false);
+    }
 
-  if (!ballot) return null;
+    if (ballot && !ballot.published) {
+      void router.push("/ballot");
```
Comment (0xmad) [inline src/pages/ballot/confirmation.tsx:20]: Why do you need void here?

File: src/features/ballot/components/BallotConfirmation.tsx (hunk 2 above)
Comment (kittybest) [reply src/features/ballot/components/BallotConfirmation.tsx]: for updating feedback link, I opened another issue for it: https://github.com/privacy-scaling-explorations/maci-rpgf/issues/166

File: src/components/ui/Loading.tsx
Comment (kittybest) [reply src/components/ui/Loading.tsx:1]: I thought I removed it, seems like not? let me try again

File: src/pages/ballot/confirmation.tsx (hunk 14 above)
Comment (kittybest) [reply src/pages/ballot/confirmation.tsx:20]: since the `router.push` function is actually an async function, so I just copy paste the code here: https://github.com/privacy-scaling-explorations/maci-rpgf/blob/da5b785c14fb0fb6ffb4bd34c47f8939aa944d8c/src/layouts/BaseLayout.tsx#L51

should both be modified by removing that `void`?

Comment (0xmad) [reply src/pages/ballot/confirmation.tsx:20]: Yes, let's use await instead.

**Review by 0xmad — APPROVED** [review]
@kittybest thanks!
